File: src/main/scala/ex6/Sortieren.scala
abstract class Sortieren(arr: Array[Int]) {

  protected val a: Array[Int] = arr.clone()
  protected val size: Int = a.length
  protected var count: Int = 0

  def sort(): Unit

  def countAndReset(): Unit = {
    println(s"Anzahl Elementvergleiche: $count")
    count = 0
  }

  def get(idx: Int): Int = {
    count += 1
    a(idx)
  }

  def swap(i: Int, j: Int): Unit = {
    val t = a(i)
    a(i) = a(j)
    a(j) = t
  }

  def rawGet(index: Int): Int = a(index)
  def rawSet(index: Int, value: Int): Unit = a(index) = value

  override def toString: String = {
    a.mkString("[", ", ", "]")
  }
}

File: src/main/cpp/ex6/Sortieren.cpp
#include "Sortieren.hpp"
#include <algorithm>

namespace ex6
{

Sortieren::Sortieren(std::span<int> arr)
    : count{0}, a{std::make_unique<int[]>(arr.size())}, size{arr.size()}
{
  std::copy(arr.begin(), arr.end(), a.get());
}

void Sortieren::countAndReset()
{
  std::cout << "Anzahl Elementvergleiche: " << count << '\n';
  count = 0;
}

bool Sortieren::less(int lhs, int rhs)
{
  count++;
  return lhs < rhs;
}

void Sortieren::swap(int i, int j)
{
  int t = a[i];
  a[i] = a[j];
  a[j] = t;
}

int &Sortieren::operator[](int index) { return a[index]; }

std::ostream &operator<<(std::ostream &os, const ex6::Sortieren &s)
{
  os << "[";
  for (size_t i = 0; i < s.size; ++i)
  {
    os << s.a[i];
    if (i < s.size - 1)
      os << ", ";
  }
  os << "]";
  return os;
}

} // namespace ex6
